### PR TITLE
feat(core): add operation state

### PR DESCRIPTION
**Pull request by griest024**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/develop/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the new behavior?
Adds state, adapter, selectors, and facade for a basic operation state. This can be used as the basis for most driver-oriented state packages, significantly DRYing up the codebase.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

### CODE REVIEW

File: libs/core/state/src/operation/facade.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+   * Whether the operation state has any errors.
+   * If so, it should be considered to be in an "error" state.
+   */
+  hasErrors$: Observable<boolean>;
+}
+
+/**
+ * @inheritdoc
+ */
+export abstract class DaffOperationStateFacade<
```
Comment (damienwebdev) [inline libs/core/state/src/operation/facade.ts:48]: The abstract class here bloats the API, imo.

The interface itself is sound, but the rest of the code exists solely to reduce code but at the cost of encouraging inheritance, which I think will make usages more brittle.

I think we keep the interface, but delete the abstract class.

File: libs/core/state/src/operation/selectors.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+   */
+  selectHasErrors: MemoizedSelector<TRootState, boolean>;
+}
+
+/**
+ * Creates a set of selectors for an operation state.
+ *
+ * @param selectState The feature selector for the operation state.
+ */
+export const daffOperationStateSelectorFactory = <
```
Comment (damienwebdev) [inline libs/core/state/src/operation/selectors.ts:49]: This is a dev time function only, it only provides organizational use to us. Is there a way to leverage the compiler's code folding feature to remove this function before runtime?

File: libs/core/state/src/operation/state.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+   */
+  loading: DaffState;
+
+  /**
+   * The errors generated by the most recent operation.
+   */
+  errors: DaffStateError[];
+}
+
+export const daffOperationInitialState: DaffOperationState = {
```
Comment (damienwebdev) [inline libs/core/state/src/operation/state.ts:20]: Same idea here with code folding, this is purely organizational code.

**Review by damienwebdev — CHANGES_REQUESTED** [review]
I'm convinced that the DRYing that this PR does is good, but I'm not convinced in the implementation.

I have a feeling that we should be able to abuse code folding https://angular.io/guide/aot-compiler#code-folding to be able to limit the runtime performance impact of some of this code.

File: libs/core/state/src/operation/selectors.ts (hunk 2 above)
Comment (griest024) [reply libs/core/state/src/operation/selectors.ts:49]: the selector factory is the entrypoint into the selector portion of this feature. Are you suggesting that I memoize this factory and only export the memoized version?

File: libs/core/state/src/operation/state.ts (hunk 3 above)
Comment (griest024) [reply libs/core/state/src/operation/state.ts:20]: this is used in tests and does have a use case in consuming code. If I were to unexport this, I would just delete the token entirely and inline the code manually.

File: libs/core/state/src/operation/adapter.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
+import { DaffState } from '../states/public_api';
+import { DaffOperationState } from './state';
+
+/**
+ * An adapter for modifying {@link DaffOperationState}.
+ */
+export class DaffOperationStateAdapter<T extends DaffOperationState = DaffOperationState> {
```
[the remark this answers is not in the document]
Comment (griest024) [reply libs/core/state/src/operation/adapter.ts]: agreed, removed

File: libs/core/state/src/operation/facade.ts (hunk 1 above)
Comment (griest024) [reply libs/core/state/src/operation/facade.ts]: agreed, removed

File: libs/core/state/src/operation/selectors.ts (hunk 2 above)
Comment (griest024) [reply libs/core/state/src/operation/selectors.ts]: changed to named function

**Pull request merged**